### PR TITLE
Add prediction early stopping

### PR DESCRIPTION
**Pull request by cbecker**
This PR adds prediction early stopping to the GBDT class, through a new prediction function `GBDT::PredictRawEarlyStop()`.

This is only one of multiple ways to implement this. Feedback is welcome. See my comments inline.

### CODE REVIEW

File: include/LightGBM/boosting.h [hunk 1]
```diff
@@ -135,6 +136,17 @@ class LIGHTGBM_EXPORT Boosting {
...
   /*!
+  * \brief Prediction for one record, similar to PredictRaw but applies early stopping to speed up prediction.
+  *     In contrast to PredictRaw(), this function runs on a single thread.
+  *
+  * \param feature_values Feature value on this record
+  * \param output Prediction result for this record
+  * \param earlyStop Early stopping instance
+  */
+  virtual void PredictRawEarlyStop(const double* features, double* output,
+                            const PredictionEarlyStopInstance& earlyStop) const = 0;
```
Comment (cbecker) [inline include/LightGBM/boosting.h]: an early stopping struct instance is needed to predict a sample, and separates prediction early stopping from the configuration of GBDT, which I think helps keep the code an functionality separate.

File: include/LightGBM/prediction_early_stop.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+  struct PredictionEarlyStopConfig
+  {
+    int roundPeriod;
+    double marginThreshold;
+  };
+
+  /// Create an early stopping algorithm of type `type`, with given roundPeriod and margin threshold
+  LIGHTGBM_EXPORT PredictionEarlyStopInstance createPredictionEarlyStopInstance(const std::string& type,
+                                                                const PredictionEarlyStopConfig& config);
```
Comment (cbecker) [inline include/LightGBM/prediction_early_stop.h]: Possible types are `none`, `multiclass` and `binary`.

File: src/boosting/prediction_early_stop.cpp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    const double marginThreshold = config.marginThreshold;
+
+    return PredictionEarlyStopInstance{
+      [marginThreshold](const double* pred, int sz)
+      {
+        // copy and sort
+        std::vector<double> votes(static_cast<size_t>(sz));
+        for (int i=0; i < sz; ++i)
+           votes[i] = pred[i];
+        std::partial_sort(votes.begin(), votes.begin() + 2, votes.end(), std::greater<double>());
```
Comment (cbecker) [inline src/boosting/prediction_early_stop.cpp]: Right now we're not verifying that the prediction vector is at least of size 2, @guolinke would it be ok to check for it here and throw an exception if this is not met?

This means that `PredictRawEarlyStop` would throw.

**Comment by guolinke** [comment]
@cbecker 
This is not a accuracy prediction, right? I mean it may drop the prediction accuracy.

**Comment by cbecker** [comment]
> This is not a accuracy prediction, right? I mean it may drop the prediction accuracy.

Yes, indeed. It's something that is useful when speed is a concern. Compared to limiting the numbe rof trees used for prediction, using this type of early stopping  'adapts' automatically to the training instance.

**Comment by guolinke** [comment]
OK. BTW, do we really need to expose the callback. Or just letting user to set the ```roundPeriod``` and ```marginThreshold``` is enough ?

**Comment by cbecker** [comment]
> OK. BTW, do we really need to expose the callback? Or just letting user to set the roundPeriod and marginThreshold is enough ?

I thought that exposing the callback is useful if the user wants to have a custom asymmetric early stopping method (e.g. if class 0 is background and 1 foreground, many place a threshold on the negative class score, and leave the positive one without early stopping. E.g. stop if binary score < -1.0)

**Comment by cbecker** [comment]
Also, do you know why the tests may be failing? There is an undefined reference to `PredictRawEarlyStop` but it looks strange to me, and it compiles successfully on my desktop (linux, gcc)

```
[0m[ 51%] [32mBuilding CXX object CMakeFiles/lightgbm.dir/src/boosting/gbdt_prediction.cpp.o
[0m[31m[1mLinking CXX executable ../lightgbm
[0mCMakeFiles/lightgbm.dir/src/boosting/gbdt.cpp.o:(.data.rel.ro._ZTVN8LightGBM4GBDTE[_ZTVN8LightGBM4GBDTE]+0xa0): undefined reference to `LightGBM::GBDT::PredictRawEarlyStop(double const*, double*, LightGBM::PredictionEarlyStopInstance const&) const'
CMakeFiles/lightgbm.dir/src/boosting/boosting.cpp.o:(.data.rel.ro._ZTVN8LightGBM4DARTE[_ZTVN8LightGBM4DARTE]+0xa0): undefined reference to `LightGBM::GBDT::PredictRawEarlyStop(double const*, double*, LightGBM::PredictionEarlyStopInstance const&) const'
CMakeFiles/lightgbm.dir/src/boosting/boosting.cpp.o:(.data.rel.ro._ZTVN8LightGBM4GOSSE[_ZTVN8LightGBM4GOSSE]+0xa0): undefined reference to `LightGBM::GBDT::PredictRawEarlyStop(double const*, double*, LightGBM::PredictionEarlyStopInstance const&) const'
collect2: error: ld returned 1 exit status
make[2]: *** [../lightgbm] Error 1
```

**Comment by guolinke** [comment]
@cbecker refer to this pr #482 . It will replace the gbdt_prediction.cpp file. I think you may need to write your prediction code in a new cpp file.

**Comment by cbecker** [comment]
> @cbecker refer to this pr #482 . It will replace the gbdt_prediction.cpp file. I think you may need to write your prediction code in a new cpp file.

Thanks, that looks a bit tricky to me, the fact that it's overwriting its own code. I'll put it in a separate file, at least for now it will do, but it won't be able to deal with pre-compiled models. Btw, which speed up do you achieve by having the model in if/else c++ statements?

**Comment by wxchan** [comment]
I think you can put your code in gbdt.cpp.

**Comment by cbecker** [comment]
> I think you can put your code in gbdt.cpp.

For me there or in the current file works too. Let me know what you prefer and I'll make the changes and rename the commit, if we can merge :+1:

**Comment by cbecker** [comment]
Done now :)

**Comment by cbecker** [comment]
> @cbecker It seems this features cannot be easy to use. User must write his own cpp code to call this function ?

True, I will add the respective C api tomorrow and get back to you.

**Comment by cbecker** [comment]
@guolinke would you agree with removing the parallel loops in PredictRaw() and Predict() now? Because then the code can be modularized and kept much simpler, and we'll have most of the prediction codein single function instead of 3 of them (including early stopping.

**Comment by guolinke** [comment]
@cbecker 
OK, you can make them to single thread.

**Comment by guolinke** [comment]
@cbecker 
BTW, can you also support this for the CLI version.

**Comment by cbecker** [comment]
Tests won't pass yet, I have to modify `ModelToIfElse()` first, but the overall functionality is there now.

**Comment by cbecker** [comment]
@guolinke I am wondering why the new two C apis are not being exported to the .so file, have you seen this before? Are you stripping symbols at some stage, and if so, how do you control which ones to keep?

File: src/c_api.cpp [hunk 4]
```diff
@@ -162,6 +163,7 @@ class Booster {
 
   void Predict(int num_iteration, int predict_type, int nrow,
                std::function<std::vector<std::pair<int, double>>(int row_idx)> get_row_fun,
+               PredictionEarlyStoppingHandle early_stop_handle,
```
Comment (guolinke) [inline src/c_api.cpp]: I think we cannot use function_pointer in c_api. It is hard to use this in python/R

Comment (guolinke) [inline src/c_api.cpp]: Oh, I am wrong. it seems you use handle. So it is a class.

**Comment by guolinke** [comment]
@cbecker 
https://github.com/cbecker/LightGBM/blob/4e3fadb31c0e1213f57b8661456fbc69bf8df5de/src/c_api.cpp#L239-L261

I think you should these two apis outside of this class. 
And they also need to be outside of namespace LightGBM.

**Comment by guolinke** [comment]
@wxchan 
can you help to review the changes of python-package and if-else part ?

**Comment by cbecker** [comment]
Before reviewing, let me fix the code so that travis is green.

File: include/LightGBM/c_api.h [hunk 5]
```diff
@@ -575,6 +577,7 @@ LIGHTGBM_C_EXPORT int LGBM_BoosterPredictForCSR(BoosterHandle handle,
                                                 int64_t num_col,
                                                 int predict_type,
                                                 int num_iteration,
+                                                PredictionEarlyStoppingHandle early_stop_handle,
```
Comment (guolinke) [inline include/LightGBM/c_api.h]: const here?

File: include/LightGBM/c_api.h [hunk 6]
```diff
@@ -612,6 +616,7 @@ LIGHTGBM_C_EXPORT int LGBM_BoosterPredictForCSC(BoosterHandle handle,
                                                 int64_t num_row,
                                                 int predict_type,
                                                 int num_iteration,
+                                                PredictionEarlyStoppingHandle early_stop_handle,
```
Comment (guolinke) [inline include/LightGBM/c_api.h]: const here?

File: include/LightGBM/c_api.h [hunk 7]
```diff
@@ -643,6 +649,7 @@ LIGHTGBM_C_EXPORT int LGBM_BoosterPredictForMat(BoosterHandle handle,
                                                 int is_row_major,
                                                 int predict_type,
                                                 int num_iteration,
+                                                PredictionEarlyStoppingHandle early_stop_handle,
```
Comment (guolinke) [inline include/LightGBM/c_api.h]: const here?

File: python-package/lightgbm/basic.py [hunk 8]
```diff
@@ -1568,7 +1581,8 @@ def dump_model(self, num_iteration=-1):
                 ptr_string_buffer))
         return json.loads(string_buffer.value.decode())
 
-    def predict(self, data, num_iteration=-1, raw_score=False, pred_leaf=False, data_has_header=False, is_reshape=True):
+    def predict(self, data, num_iteration=-1, raw_score=False, pred_leaf=False, data_has_header=False, is_reshape=True,
+                early_stop_instance=None):
```
Comment (guolinke) [inline python-package/lightgbm/basic.py]: why new line?

File: src/boosting/gbdt_prediction.cpp [hunk 9]
```diff
@@ -17,44 +18,33 @@
...
-    for (int k = 0; k < num_tree_per_iteration_; ++k) {
-      double t = 0.0f;
-      #pragma omp parallel for schedule(static) reduction(+:t)
-      for (int i = 0; i < num_iteration_for_pred_; ++i) {
-        t += models_[i * num_tree_per_iteration_ + k]->Predict(features);
-      }
-      output[k] = t;
-    }
+void GBDT::PredictRaw(const double* features, double* output, const PredictionEarlyStopInstance* earlyStop) const {
+  const auto noEarlyStop = createPredictionEarlyStopInstance("none", PredictionEarlyStopConfig());
```
Comment (guolinke) [inline src/boosting/gbdt_prediction.cpp]: The ``PredictRaw`` will be called one time for one instance. I am afraid of the efficiency here.

**Comment by wxchan** [comment]
@cbecker can you also add some python test case? It's better to have both binary and multiclass cases.

File: python-package/lightgbm/basic.py [hunk 10]
```diff
@@ -409,7 +418,7 @@ def predict(self, data, num_iteration=-1,
                                                predict_type)
         elif isinstance(data, DataFrame):
             preds, nrow = self.__pred_for_np2d(data.values, num_iteration,
-                                               predict_type)
+                                               predict_type, early_stop_instance_handle)
```
Comment (wxchan) [inline python-package/lightgbm/basic.py]: is this `early_stop_instance_handle` a typo? seem only shown once

File: src/boosting/gbdt_prediction.cpp [hunk 11]
```diff
@@ -17,44 +18,33 @@
...
-      for (int i = 0; i < num_iteration_for_pred_; ++i) {
-        t += models_[i * num_tree_per_iteration_ + k]->Predict(features);
-      }
-      output[k] = t;
+
+    // check early stopping
+    ++earlyStopRoundCounter;
+    if (earlyStop->roundPeriod == earlyStopRoundCounter) {
+      if (earlyStop->callbackFunction(output, num_tree_per_iteration_))
+        return;
```
Comment (guolinke) [inline src/boosting/gbdt_prediction.cpp]: add ```{... }```.

**Comment by wxchan** [comment]
Is there any introduce on early stopping in prediction? Sorry I never heard about this before.

File: src/boosting/gbdt_prediction.cpp [hunk 12]
```diff
@@ -17,44 +18,33 @@
...
+      earlyStopRoundCounter = 0;
     }
   }
+}
+
+void GBDT::Predict(const double* features, double* output, const PredictionEarlyStopInstance* earlyStop) const {
+  PredictRaw(features, output, earlyStop);
```
Comment (guolinke) [inline src/boosting/gbdt_prediction.cpp]: ```PredictRaw``` is not inline. maybe change it to inline is better.

File: src/boosting/prediction_early_stop.cpp [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+    // marginThreshold will be captured by value
+    const double marginThreshold = config.marginThreshold;
+
+    return PredictionEarlyStopInstance{
+      [marginThreshold](const double* pred, int sz)
+      {
+        // copy and sort
+        std::vector<double> votes(static_cast<size_t>(sz));
+        for (int i=0; i < sz; ++i)
+           votes[i] = pred[i];
```
Comment (guolinke) [inline src/boosting/prediction_early_stop.cpp]: ```{ ... } ```

File: src/boosting/prediction_early_stop.cpp (hunk 3 above)
Comment (guolinke) [inline src/boosting/prediction_early_stop.cpp]: you can just use the margin of binary class when ```votes.size() == 1```

**Comment by guolinke** [comment]
@cbecker 
Can we also support this in CLI version ? By passing the parameters when running LightGBM.
And we can create an singleton Handle in CLI version, So that we don't need to create the instance of handle every time calling the prediction (https://github.com/cbecker/LightGBM/blob/5f50904dceec524451e17bae1ac7308ab180fc50/src/boosting/gbdt_prediction.cpp#L22-L26) .

**Comment by wxchan** [comment]
The if-else part won't be a problem as long as it passes the test.

**Comment by cbecker** [comment]
@wxchan This paper shows a few methods for early stopping: https://papers.nips.cc/paper/3265-multiple-instance-pruning-for-learning-efficient-cascade-detectors.pdf

Most of them are more sophisticated than what we implement here, but the overall idea is typically similar.

File: python-package/lightgbm/basic.py (hunk 10 above)
Comment (cbecker) [inline python-package/lightgbm/basic.py]: this was fixed in the new version, the python code was not ready yet at the time you did the review

File: include/LightGBM/c_api.h (hunk 5 above)
Comment (cbecker) [inline include/LightGBM/c_api.h]: thanks, fixed.

File: include/LightGBM/c_api.h (hunk 6 above)
Comment (cbecker) [inline include/LightGBM/c_api.h]: thanks, fixed.

File: include/LightGBM/c_api.h (hunk 7 above)
Comment (cbecker) [inline include/LightGBM/c_api.h]: thanks, fixed.

File: python-package/lightgbm/basic.py (hunk 8 above)
Comment (cbecker) [inline python-package/lightgbm/basic.py]: I thought it would help readability.

File: src/boosting/gbdt_prediction.cpp (hunk 11 above)
Comment (cbecker) [inline src/boosting/gbdt_prediction.cpp]: thanks, fixed.

File: src/boosting/gbdt_prediction.cpp (hunk 12 above)
Comment (cbecker) [inline src/boosting/gbdt_prediction.cpp]: I will try. Anyway this call is probably not much compared to predicting each individual tree.

File: src/boosting/prediction_early_stop.cpp (hunk 13 above)
Comment (cbecker) [inline src/boosting/prediction_early_stop.cpp]: About using the margin of the binary class, that is why we have a specialization for binary classification.

I've just fixed the {}.

Comment (guolinke) [inline src/boosting/prediction_early_stop.cpp]: ok. you can add a ```CHECK(sz>1);``` here.

File: python-package/lightgbm/__init__.py [hunk 14]
```diff
@@ -27,4 +27,4 @@
            'train', 'cv',
            'LGBMModel', 'LGBMRegressor', 'LGBMClassifier', 'LGBMRanker',
            'print_evaluation', 'record_evaluation', 'reset_parameter', 'early_stopping',
-           'plot_importance', 'plot_metric', 'plot_tree', 'create_tree_digraph']
+           'plot_importance', 'plot_metric', 'plot_tree', 'create_tree_digraph', 'PredictionEarlyStopInstance']
```
Comment (wxchan) [inline python-package/lightgbm/__init__.py]: it's better to put this in the same line as `Dataset` and `Booster`, line 26

File: src/boosting/gbdt_prediction.cpp (hunk 9 above)
Comment (cbecker) [inline src/boosting/gbdt_prediction.cpp]: A singleton would do here, so there's no need to construct it every time, I'm on it.

File: python-package/lightgbm/basic.py [hunk 15]
```diff
@@ -1785,3 +1800,32 @@ def set_attr(self, **kwargs):
...
+class PredictionEarlyStopInstance(object):
+    """"PredictionEarlyStopInstance in LightGBM."""
+    def __init__(self, early_stop_type="none", round_period=20, margin_threshold=1.5):
+        """
+        Create an early stopping object
+
+        Parameters
+        ----------
+        early_stop_type: string
+            "none", "binary" or "multiclass". Regression is not supported.
```
Comment (wxchan) [inline python-package/lightgbm/basic.py]: I think you can just use `None` instead of string 'none' here

Comment (wxchan) [inline python-package/lightgbm/basic.py]: Never mind, seem "none" is better

File: python-package/lightgbm/__init__.py (hunk 14 above)
Comment (cbecker) [inline python-package/lightgbm/__init__.py]: Thanks, done.

File: python-package/lightgbm/basic.py (hunk 15 above)
Comment (cbecker) [inline python-package/lightgbm/basic.py]: Thanks, I allow for both options now.

**Comment by cbecker** [comment]
> @guolinke Can we also support this in CLI version ? By passing the parameters when running LightGBM. And we can create an singleton Handle in CLI version, So that we don't need to create the instance of handle every time calling the prediction (https://github.com/cbecker/LightGBM/blob/5f50904dceec524451e17bae1ac7308ab180fc50/src/boosting/gbdt_prediction.cpp#L22-L26) .

I am not sure where to modify the code for that. I think it'd be safer if you do that, as I am not familiar with the CLI code. Let me know if you agree with the state of the current PR.

On my side I am getting a 2x speed up with very little loss in accuracy in one of my tests. I am classifying millions of samples, and I got down from 1 hour to 1/2 an hour :)

**Comment by wxchan** [comment]
I think it's ready to merge for python and if-else part. You can add an example in examples/python-guide/advanced_example.py if you want.

**Comment by wxchan** [comment]
by the way, you need to update `windows` folder for new files. it's easy to forget.

**Comment by guolinke** [comment]
@cbecker OK, i can do it. 
BTW, I update your code directly for the parallel by samples.

**Comment by cbecker** [comment]
> by  the way, you need to update windows folder for new files. it's easy to forget.

I'm not sure what this means, is it something I have to do?

**Comment by guolinke** [comment]
@cbecker 
I think wxchan's point is update the ```lightgbm.vcproj``` in the ```windows``` folder.

**Comment by wxchan** [comment]
yes, LightGBM.vcxproj and LightGBM.vcxproj.filters

**Comment by cbecker** [comment]
> yes, LightGBM.vcxproj and LightGBM.vcxproj.filters

Thanks, I edited the file and made the changes manually as I don't have MSVC here, let me know if there are any issues.

**Comment by guolinke** [comment]
@wxchan Do you have any other comments ?

**Comment by wxchan** [comment]
@guolinke no.

**Comment by ddDragon** [comment]
I used the early stopping parameter about a month ago, I found that didn't work? so this version is fixed the issue? or May be the method that i used was wrong?

**Comment by guolinke** [comment]
@ddDragon This is the early stopping for the prediction, not the training. 

And I think the early stopping for training is always working.  Maybe you use it by the wrong method.

**Comment by ddDragon** [comment]
I used the parameters in lgb.train api named early_stopping_rounds. I set 30 rounds early stopping, total 5000 iterations training. I found the best valid set score was at about 2000 rounds. But all times the model trained all the 5000 rounds? This confused me a lot.

**Comment by guolinke** [comment]
@wxchan any idea about  that ? 
@ddDragon can you provide the re-produce script ?

**Comment by ddDragon** [comment]
So sorry that I did't backup my code. Maybe something wrong in my code. If I meet this problem next time I would save it. Lgb is mush faster than xgb without loss accuracy. Very nice tool!

**Pull request merged**